### ISSUE TITLE
[#88] Phase E — Julia cube-cavity Helmholtz reference (LinearAlgebra + Arpack.jl)

### ISSUE THREAD

**Issue by rjwalters** [rjwalters/geode-fem#115]
Epic: #88, Phase E.

Port the cube-cavity scalar Helmholtz reference to Julia, mirroring the structure
of the NumPy reference (#92 / PR #98) and the JAX + TF-Java pair (#93 / PR #97).
Goal: the Julia output joins the existing reference set under the same
golden-fixture schema, with eigenvalue agreement against the cube-cavity baseline.

Per Epic #88, Julia is included for:
- Native complex types throughout (vs the f64-pair representation used in Burn/NumPy)
- Mature complex-symmetric sparse eigensolvers (`Arpack.jl`)
- Probing whether complex-arithmetic ergonomics surface different L4 friction
  than what surfaces in NumPy/Burn

## Anchor references
- Epic: #88 (full scope, acceptance, principles)
- NumPy reference (pattern to mirror): #92 / PR #98
- JAX + TF-Java pair (CI-gating pattern to mirror): #93 / PR #97
- Reference harness scaffolding: #89 / PR #94
- Likely workspace location: `crates/geode-validation/reference/julia/` (parallel
  to `reference/numpy/`, `reference/jax/`, `reference/tf_java/`)

## Notes for the curator
- Confirm with the existing CI gate pattern (`.github/workflows/`) how the Julia
  job should be sequenced — by analogy to the TF-Java CI gate added in PR #102
  / renamed in PR #112 (#111).
- The Phase E acceptance bar is the same cube-cavity eigenvalue/eigenvector
  comparison that NumPy + JAX + TF-Java are already passing — golden values
  live with the fixture; tolerance per the cross-platform tolerance work
  in PR #110 / #113.
- Sparse eigensolver choice: `Arpack.jl` per the epic; document if a different
  Julia sparse eigensolver is preferable.

Filing as a raw shell off Epic #88's Phase E; curator should add the full
acceptance criteria, file-layout sketch, and CI sequencing notes.

**Comment by rjwalters** [comment]
## Curator pass 2026-06-04

Curator note: enriched per the standard L4 spine slice pattern established by NumPy (#92 / PR #98), JAX (#93 / PR #97), and TF-Java (#102 / PR #107, rename in #111 / PR #112). All file paths and CI sequencing notes below are anchored against `origin/main` at commit `acb7be7`.

### Scope

Port the cube-cavity scalar Helmholtz pipeline (mesh → P1 assembly → Dirichlet → generalized eigensolve) to Julia, mirroring the structure of the sibling NumPy / JAX / TF-Java references, and add Julia as a fifth row to the cross-IR eigenvalue agreement gate. The Julia output joins the existing reference set under `schema_version: "1"` (see `reference/SCHEMA.md`) and the cube-cavity per-fixture extension (see `reference/fixtures/cube_cavity/baseline.schema.md`).

Julia is the **complex-arithmetic** reference per Epic #88's principle 4 (the L4 calculus is interpreted by what produces equivalent output across backends). Even though the cube-cavity slice is real-symmetric, landing Julia now lays the toolchain track for Phases G–J where complex types matter (Nédélec, PML, NLEPS), and surfaces complex-arithmetic ergonomics friction earlier rather than later.

### File layout sketch (mirrors NumPy / JAX, parallel to TF-Java's Maven seam)

```
reference/julia/
├── README.md                              — updated from current stub
├── Project.toml                           — pinned deps (Arpack, MAT or JSON3, SparseArrays, LinearAlgebra)
├── Manifest.toml                          — reproducible lock (gitignored on first pass; revisit per #88 OQ#1)
├── cube_cavity.jl                         — pipeline: mesh I/O → P1 assembly → Dirichlet → Arpack.eigs
├── gen_cube_cavity_baseline.jl            — emits reference/fixtures/cube_cavity/julia_baseline.json
│                                            (parallel to gen_cube_cavity_baseline.py and
│                                             gen_cube_cavity_fixture.py)
└── mesh.jl                                — local equivalent of reference/numpy/mesh.py
                                              (cube_tet_mesh, load_msh, cube_interior_mask)

reference/fixtures/cube_cavity/
└── julia_baseline.json                    — NEW; schema v1; lowest 5 eigenvalues
                                              + k_int_frobenius / m_int_frobenius / k_int_diag / m_int_diag
                                              + n_int (same field set as baseline.json + jax_baseline.json)

reference/driver/
└── compare_eigenvalues.py                 — EDIT: add `--julia path/to/julia_baseline.json` flag
                                              parallel to existing --tfjava / --jax / --numpy / --burn.
                                              Update title rendering to handle Julia presence/absence
                                              (same pattern PR #112 used to scrub Burn placeholder).

.github/workflows/
└── julia-cube-cavity.yml                  — NEW; gated on path filter for reference/julia/** + the
                                              julia_baseline.json + the comparator. Pattern mirrors
                                              .github/workflows/tfjava-cube-cavity.yml.

crates/geode-validation/tests/
└── cube_cavity_julia_reference.rs         — NEW; loads julia_baseline.json via Fixture::load and
                                              compares Burn against it. Mirrors cube_cavity_jax_reference.rs.
                                              Same dual ndarray-f64 / wgpu-f32 tolerance split as the
                                              NumPy/JAX comparators (NDARRAY_F64_TOLERANCES + GPU_F32_TOLERANCES).
```

### Acceptance criteria

**AC#1 — Pipeline end-to-end in Julia.** `reference/julia/cube_cavity.jl` runs the full pipeline on the canonical `unit_cube.msh` (n=10 tet-split cube, MSH 4.1 ASCII, the same file that the NumPy + JAX + TF-Java references consume) and produces the lowest-5 generalized eigenvalues of (K_int, M_int).

**AC#2 — Eigenvalue agreement with the existing reference set.** Julia eigenvalues agree with the NumPy `baseline.json` and JAX `jax_baseline.json` rows on the lowest 5 modes to **`1e-5` relative** (same headline tolerance the TF-Java gate enforces in `.github/workflows/tfjava-cube-cavity.yml`). This is the cross-IR f64 agreement floor per the comparator's defaults — tighter than the f32 GPU friction floor (1e-7) but loose enough to absorb ARPACK-vs-ARPACK iteration-trace differences.

**AC#3 — Sub-stage f64 agreement matches the cross-platform floor.** `k_int_frobenius`, `m_int_frobenius`, `k_int_diag`, `m_int_diag` agree with the values pinned in `reference/fixtures/cube_cavity/baseline.json` to the same tolerances PR #113 (issue #110) calibrated cross-platform:

| Field | Tolerance |
|---|---|
| `k_int_frobenius` | `1e-9` abs |
| `m_int_frobenius` | `1e-8` abs |
| `k_int_diag` | `1e-9` abs |
| `m_int_diag` | `5e-9` abs |

This is the same floor that absorbed the rustc/LLVM SIMD-reduction-order variance across Ubuntu / macOS arm64 / macOS Intel. Julia's BLAS (OpenBLAS by default on most platforms) has its own reduction-order quirks; the curated expectation is that Julia lands inside the same `5e-9` envelope, but **if Julia exceeds the floor on any of the three CI matrix platforms, that is a Friction Artifact** (file on #5 per Epic #88's friction-mining loop), not a license to loosen the floor.

**AC#4 — Julia baseline fixture lands in canonical schema.** `reference/fixtures/cube_cavity/julia_baseline.json` is `schema_version: "1"`, has the same input/output field set as `jax_baseline.json`, and is consumed by the new Rust comparator test via `geode-validation::Fixture::load`. Regeneration is deterministic on a pinned `Arpack.jl` (`Project.toml` pins the version).

**AC#5 — CI gate on `julia-cube-cavity.yml`.** New workflow mirroring `.github/workflows/tfjava-cube-cavity.yml`:

- Path-filtered triggers: `reference/julia/**`, `reference/driver/compare_eigenvalues.py`, `reference/numpy/cube_cavity_minimal.py`, `reference/numpy/requirements.txt`, `reference/fixtures/cube_cavity/jax_baseline.json`, `reference/fixtures/cube_cavity/baseline.json`, `.github/workflows/julia-cube-cavity.yml`.
- `workflow_dispatch` always available with optional `n` and `rtol` inputs (defaults `n=10`, `rtol=1e-5` — matching the cube-cavity baseline mesh, not the TF-Java gate's n=4 cost-floor compromise).
- Step sequence: `julia-actions/setup-julia` → `Pkg.instantiate()` → run `cube_cavity.jl` → emit Julia sidecar JSON → invoke `compare_eigenvalues.py --julia ... --jax ... --numpy ... --rtol 1e-5`.
- Hard exit-1 on disagreement; upload agreement table as artifact on every run (pass or fail), same as the TF-Java gate.

**AC#6 — Rust comparator test.** `crates/geode-validation/tests/cube_cavity_julia_reference.rs` loads `julia_baseline.json` and asserts agreement against the Burn cube-cavity path. Same backend-aware tolerance pattern as `cube_cavity_jax_reference.rs`:

```rust
// f64 ndarray backend → tight tolerance
// wgpu/cuda backend → looser f32 tolerance per GPU_F32_TOLERANCES
```

Test is `#[ignore]` by default (same rationale as the existing sibling tests — faer's `qz_real` panics under debug-assertions; runs only with `cargo test --release -- --ignored`).

**AC#7 — `reference/julia/README.md` upgraded** from the current stub to document:
- Toolchain bootstrap (`julia --project=. -e 'using Pkg; Pkg.instantiate()'`).
- Invocation convention (`julia --project=. cube_cavity.jl <inputs> <outputs>` — same subprocess pattern the NumPy README documents).
- The complex-arithmetic friction observation slot (initially empty — populated once concrete frictions surface; this slice is real-symmetric so the slot may stay empty until Phase G's Nédélec).
- ARPACK iteration-trace caveat: re-runs against a different `Arpack.jl` version may produce eigenvectors differing by orthogonal rotation within degenerate clusters; the **eigenvalues** are stable, the **eigenvectors** require the subspace-overlap convention (see `baseline.schema.md` "Per-cluster subspace-overlap convention"). The Julia fixture stores eigenvalues only (mirroring `jax_baseline.json`), not eigenvectors — sufficient for AC#2 / AC#5 / AC#6.

### CI sequencing notes

The new `.github/workflows/julia-cube-cavity.yml` should be modeled after `.github/workflows/tfjava-cube-cavity.yml` rather than `arpack.yml`, because the TF-Java gate is the **canonical sidecar reference pattern** (own toolchain, own JSON sidecar, gated path filter, hard agreement gate, artifact upload). Specifically:

1. **Toolchain isolation.** Use `julia-actions/setup-julia@v2` with a pinned Julia version (suggested: `'1.10'` LTS line; revisit if `Arpack.jl` drops 1.10 support). Cache the Julia depot via `julia-actions/cache@v2` keyed off `Project.toml`. Do **not** install JAX, do not install Burn — Julia is its own toolchain seam, exactly like TF-Java is its own JVM seam.

2. **Reuse the JAX baseline fixture as the cross-IR comparison anchor.** The JAX baseline at `reference/fixtures/cube_cavity/jax_baseline.json` is the in-tree anchor that the TF-Java gate already compares against. Julia compares against the *same* fixture for consistency. The NumPy row is freshly emitted in-job (per `emit_numpy_eigenvalues.py`) for the same reason the TF-Java gate does it.

3. **No Burn row in the Julia gate.** Burn agreement is exercised separately by the per-push `cube_cavity_jax_reference` cargo test and by `arpack.yml` against the same JAX baseline. PR #112 (issue #111) established this scope-clarification pattern explicitly — do not regress it for the Julia gate.

4. **Job name and step name.** Per PR #112's naming convention: `Julia assembly + Arpack.jl eigensolve + cross-IR agreement (Julia vs JAX vs NumPy)`. Gate step: `Cross-IR agreement gate (Julia vs JAX vs NumPy, rtol=...)`.

5. **n=10 on the Julia gate, vs n=4 on the TF-Java gate.** The TF-Java gate uses n=4 to keep the JVM-heavy job under reasonable wall-clock. Julia + Arpack.jl is closer to NumPy + ARPACK in cost (both go through libarpack); n=10 (matching `baseline.json`) is the natural target. If wall-clock becomes a problem, fall back to n=4 with an explicit note in the workflow header — but **start with n=10** since AC#3 references the n=10 sub-stage fixture.

### Eigensolver choice — confirm Arpack.jl

Epic #88 specifies `Arpack.jl` and this issue inherits that choice. **Curator confirms Arpack.jl** as the right pick for Phase E because:

- It is the Julia binding to the same `libarpack` that NumPy/SciPy uses under `scipy.sparse.linalg.eigsh`. Using it makes the Julia eigensolve a near-bit-equivalent of the NumPy reference at the iteration-trace level — surface a different result only if Julia's wrapping introduces drift, not because of algorithm choice. This is the **cleanest possible isolation** of the "Julia vs NumPy" friction question for the eigensolve step.
- It is mature, has clean handling for the shift-and-invert mode (`sigma=0`, `which=:LM`) that the NumPy reference relies on, and has a public API stable since 0.5.x.

**Override candidates if Arpack.jl proves problematic:**
- `KrylovKit.jl` — pure-Julia Lanczos/Arnoldi, no libarpack dependency. Would give a cleaner "Julia stack only" friction artifact but would also dilute the iteration-trace agreement signal with NumPy. Recommended only if `Arpack.jl` has install/build issues on the GitHub Actions runner.
- `ArnoldiMethod.jl` — another pure-Julia option; same trade-off as KrylovKit.

The decision is: **default to Arpack.jl; document any switch as the first Julia-specific Friction Artifact filed on #5**. Don't switch silently.

### Open questions worth surfacing

1. **`Manifest.toml` in git?** Julia best practice is to commit `Manifest.toml` for applications and **not** commit it for libraries. The Julia reference is an application (a fixture generator), so `Manifest.toml` *should* be checked in for reproducibility. However: this conflicts mildly with the cross-backend "every toolchain owns its own pinning" convention established for NumPy (which uses `requirements.txt`, not `pip-tools`-style lock files). **Recommended: commit `Manifest.toml` initially**; revisit if it produces churn. Either way, document the choice in the README.

2. **JSON I/O library.** Julia has multiple options (`JSON.jl`, `JSON3.jl`). `JSON3.jl` is fastest and supports the StructTypes ergonomics, but `JSON.jl` is more universally available. **Recommended: `JSON3.jl`** — the cube-cavity sub-stage diagonals are 729-element arrays, and JSON3's streaming serialization is more honest about the schema-v1 row-major flattening convention. Either is acceptable.

3. **Mesh I/O — `Gmsh.jl` or inline parser?** The NumPy reference uses `meshio`. Julia has `Gmsh.jl` (binding to libgmsh) which would be the parallel choice, but it pulls a ~50 MB native dependency. The mesh files we consume are MSH 4.1 ASCII — small enough to parse inline in `reference/julia/mesh.jl` without a libgmsh dep, mirroring the way NumPy's `mesh.py` doesn't actually need `meshio` for the cube case (it generates the tet split procedurally). **Recommended: inline parser for MSH 4.1 ASCII in `reference/julia/mesh.jl`**, matching the algorithmic structure of `reference/numpy/mesh.py::cube_tet_mesh`. Avoids libgmsh dep on the CI runner.

4. **f64 default.** Julia's default `Float64` is f64. **No explicit dtype flag needed** (unlike JAX, where `jax.config.update("jax_enable_x64", True)` is mandatory at the top of every script — that itself is a JAX-DX friction artifact recorded in PR #97). The fact that Julia "just is" f64 is itself an interesting L4 friction observation — **document it in the Julia README as a positive friction artifact**: "what JAX requires you to explicitly enable, Julia gives you for free."

5. **Wall-clock budget for the CI gate.** Estimated wall-clock for n=10 Julia + Arpack.jl: ~5–15 seconds for the eigensolve, ~30–60 seconds for `Pkg.instantiate()` on a cold cache, ~10–20 seconds for the actual run. Total cold-cache: ~1–2 minutes; warm-cache: ~30 seconds. Comparable to the TF-Java gate's ~3–4 minute Maven-cold wall-clock. Acceptable.

6. **Eigenvectors in the fixture?** `jax_baseline.json` ships eigenvalues + sub-stage diagnostics only, **not** eigenvectors. `baseline.json` (the NumPy canonical) ships eigenvectors as INPUTS (Q_numpy) for the subspace-overlap check. **Recommendation: Julia fixture mirrors `jax_baseline.json` shape** (eigenvalues + diagnostics only, no eigenvectors). The subspace-overlap check is anchored to NumPy as canonical tiebreaker per Epic #88; Julia agreement on eigenvalues is the headline cross-IR signal. If a follow-up wants Julia eigenvectors for an independent subspace-overlap audit, that's a separate slice.

### Anchor references (re-stated for completeness)

- Epic: #88 (full scope, principles, acceptance phase list, memory-architecture commitments)
- Reference harness scaffolding: #89 / PR #94 — establishes `reference/SCHEMA.md` v1 and `reference/fixtures/` directory convention
- NumPy cube-cavity end-to-end: #92 / PR #98 — establishes `baseline.json` schema extension and sub-stage f64 friction pattern
- JAX + TF-Java cube-cavity pair: #93 / PR #97 — establishes the per-backend pipeline file pattern and `jax_baseline.json` shape
- TF-Java CI gate: #102 / PR #107 — establishes the gated path-filtered CI workflow pattern
- TF-Java CI gate scope clarification: #111 / PR #112 — establishes three-way agreement convention (and Burn-elsewhere footnote)
- Cross-platform tolerance floor: #110 / PR #113 — establishes the `1e-9` / `5e-9` sub-stage f64 cross-platform floor
- Friction-mining loop: file disagreements on #5 with the L4 operator name and affected backend pair

### Labels (curator action)

- `loom:curated` (added by curator) — replaces `loom:triage`
- `tier:goal-supporting` retained
- Note: **not** marking `loom:issue` — human-only label per Loom convention.

### Suggested builder approach

1. **Land the Julia reference impl first** (`reference/julia/cube_cavity.jl`, `mesh.jl`, `Project.toml`) and verify it produces eigenvalues matching `baseline.json` locally.
2. **Add the fixture generator** (`gen_cube_cavity_baseline.jl`) and commit `julia_baseline.json`.
3. **Add the comparator flag** (`--julia` in `compare_eigenvalues.py`) — mechanical extension of the existing pattern.
4. **Add the Rust test** (`cube_cavity_julia_reference.rs`) — copy-paste from `cube_cavity_jax_reference.rs`, swap the fixture path.
5. **Add the CI workflow** (`julia-cube-cavity.yml`) — copy-paste from `tfjava-cube-cavity.yml`, swap the toolchain block.
6. **Upgrade the README** with toolchain bootstrap + friction observations slot.

Each of steps 1–6 is independently reviewable; PR can land them all together or split into 2–3 PRs (impl + CI gate + Rust test would be a natural split). Recommended: single PR, since the contract pieces (fixture, comparator, CI gate) are tightly coupled and reviewing them in lockstep is easier.
